grpclog/grpclog.go

```diff
@@ -0,0 +1,441 @@
+package grpclog
+
+import (
+	"encoding/json"
+	"os"
+	"path/filepath"
+	"strings"
+
+	rpc "github.com/pulumi/pulumi/sdk/v3/proto/go"
+	"google.golang.org/protobuf/types/known/emptypb"
+)
+
+type GrpcLog struct {
+	Entries []GrpcLogEntry `json:"entries"`
+}
+
+type GrpcLogEntry struct {
+	// The gRPC method being called.
+	// This is the full method name, e.g. "/pulumirpc.ResourceProvider/Check".
+	Method   string          `json:"method"`
+	Request  json.RawMessage `json:"request,omitempty"`
+	Response json.RawMessage `json:"response,omitempty"`
+}
+
+type Method string
+
+const (
+	Attach        Method = "/pulumirpc.ResourceProvider/Attach"
+	Call          Method = "/pulumirpc.ResourceProvider/Call"
+	Cancel        Method = "/pulumirpc.ResourceProvider/Cancel"
+	Check         Method = "/pulumirpc.ResourceProvider/Check"
+	CheckConfig   Method = "/pulumirpc.ResourceProvider/CheckConfig"
+	Configure     Method = "/pulumirpc.ResourceProvider/Configure"
+	Construct     Method = "/pulumirpc.ResourceProvider/Construct"
+	Create        Method = "/pulumirpc.ResourceProvider/Create"
+	Delete        Method = "/pulumirpc.ResourceProvider/Delete"
+	Diff          Method = "/pulumirpc.ResourceProvider/Diff"
+	DiffConfig    Method = "/pulumirpc.ResourceProvider/DiffConfig"
+	GetMapping    Method = "/pulumirpc.ResourceProvider/GetMapping"
+	GetMappings   Method = "/pulumirpc.ResourceProvider/GetMappings"
+	GetPluginInfo Method = "/pulumirpc.ResourceProvider/GetPluginInfo"
+	GetSchema     Method = "/pulumirpc.ResourceProvider/GetSchema"
+	Invoke        Method = "/pulumirpc.ResourceProvider/Invoke"
+	Read          Method = "/pulumirpc.ResourceProvider/Read"
+	Update        Method = "/pulumirpc.ResourceProvider/Update"
+)
+
+type AttachLog struct {
+	Request  rpc.PluginAttach
+	Response emptypb.Empty
+}
+
+func (l *GrpcLog) Attaches() ([]AttachLog, error) {
+	var typedEntries []AttachLog
+	for _, entry := range l.WhereMethod(Attach) {
+		var typedEntry AttachLog
+		err := unmarshalEntry(entry, &typedEntry.Request, &typedEntry.Response)
+		if err != nil {
+			return nil, err
+		}
+		typedEntries = append(typedEntries, typedEntry)
+	}
+	return typedEntries, nil
+}
+
+type CallLog struct {
+	Request  rpc.CallRequest
+	Response rpc.CallResponse
+}
+
+func (l *GrpcLog) Calls() ([]CallLog, error) {
```

Comment:
Could all these mostly identical reader methods be one generic method, parametrized on the return type?

Reply:
Maybe ... though I think it's useful to have the named methods for discoverability, but could perhaps use generics internally.